File: src/fit_model.R
# Fit linear and loglinear models of drift for GO term drift analysis

get_formula <- function(model_type, interaction)
{
    # Return a formula object
    # Can set a linear or loglinear model of the day_harvested column
    # Can also decide if we want + or * for checking the interaction with drug
    return (as.formula(paste0(
        "drift ~ ",
        ifelse(model_type == "linear", "day_harvested", "log(day_harvested)"),
        ifelse(interaction, " * ", " + "),
        "drug"
    )))
}

fit_model <- function(drift, model_type, interaction)
{
    # Fit a linear model for a precalculated drift dataframe.
    # Arguments:
    #   drift: a dataframe
    #     must have columns: drift, day_harvested, drug
    #   model_type: a string; one of ["linear", "loglinear"]
    #   interaction: a boolean
    #
    # Returns:
    #   A list of three elements:
    #     1. The ANOVA pvalue of the model with the interaction term
    #     2. The adjusted p value of the linear model
    #     3. The coefficient matrix of the linear model

    formula <- get_formula(model_type, interaction)
    other_formula <- get_formula(model_type, !interaction)

    model <- lm(formula, data = drift)
    interact <- lm(other_formula, data = drift)

    anova_pval <- anova(model, interact)$`Pr(>F)`[2]

    res <- summary(model)
    adj_r <- res$adj.r.squared

    return (list(anova_pval, adj_r, as.data.frame(res$coefficients)))
}

File: src/fit_model.py
"""Python wrapper for the R linear model fitter."""

import os
import rpy2.robjects as robj

def fit_model(df, model_type, interaction):
    """Fit a particular linear model to a drift dataframe.

    Params:
        df: a pandas dataframe
        model_type: one of {"linear", "log_linear"} describing whether drift
            increases linearly or log-linearly with time
        interaction: a Boolean indicating whether to check the interaction term
            between time and the drug treatment
    """

    R = robj.r
    fname = os.path.realpath(__file__).replace(".py", ".R")
    R.source(fname)

    res = R.fit_model(df, model_type, interaction)

    anova_pval = res[0][0]
    adj_rsq = res[1][0]

    return (robj.pandas2ri.ri2py(res[2]) # the coefficient dataframe
        .reset_index()
        .rename(columns = {
            "index": "parameter",
            "Estimate": "estimate",
            "Std. Error": "std_error",
            "t value": "t_value",
            "Pr(>|t|)": "p_value"
        })
        .assign(
            adj_rsq = adj_rsq,
            anova_pvalue = anova_pval,
            interaction = interaction,
            model = model_type
        )
    )
